### PR TITLE
ISO19115 basic functionalities 

### PR DESCRIPTION
**Pull request by AlexDo1**
This Pull Request adds the `standard_export` extension to metacatalog.
The extension adds a method to the Entry class which gives functionality to export the `ImmutableResultSet` of the Entry as ISO 19115.

In the extension, utility functions are included to infer needed information for ISO export from ImmutableResultSets. Furthermore, a jinja2 XML template is included, which renders the inferred information into the ISO XML file.

### CODE REVIEW

File: metacatalog/ext/standard_export/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from .extension import StandardsExportExtension
```
Comment (mmaelicke) [inline metacatalog/ext/standard_export/__init__.py:1]: The extension should be called `standards_export` as it is not the standard export extension. Also to be in line with the Class name

File: metacatalog/ext/standard_export/extension.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        template = _init_iso19115_jinja()
+
+        # render template with entry_dict
+        xml = template.render(**iso_input, **config_dict)
+
+        # check whether xml is well-formed
+        _validate_xml(xml)
+
+        # convert to ElementTree and return
+        return ET.ElementTree(ET.fromstring(xml))
```
Comment (mmaelicke) [inline metacatalog/ext/standard_export/extension.py:109]: can yon add a blank line?

File: metacatalog/ext/standard_export/extension.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                * city: str
+                * administrativeArea: str
+                * postalCode: str
+                * country: str
+                * electronicMailAddress: list(str)
+                * linkage: str
+                * linkage_name: str
+                * linkage_description: str
+            * publisher
+                * organisation_name: str
```
Comment (mmaelicke) [inline metacatalog/ext/standard_export/extension.py:80]: you need blank lines between the different sections for sphinx to recognize the list. Better would be to add a code block:

```rst
.. code-block:: Python

  dict(
    organisationName: str = '',
    deliveryPoint: str = '',
  )
```

and have a blank line before and after the block definition

File: metacatalog/ext/standard_export/extension.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        :class:`ImmutableResultSet <metacatalog.utils.results.ImmutableResultSet>`.
+        This will lazy-load sibling Entries and parent groups as needed for
+        a useful Metadata export.
+
+        """
+        # get necessary input parameters from ImmutableResultSet for ISO export
+        iso_input = _parse_iso_information(entry_or_resultset)
+
+        # get initialized jinja template
+        template = _init_iso19115_jinja()
```
Comment (mmaelicke) [inline metacatalog/ext/standard_export/extension.py:100]: The user should be able to pass in an argument to this function in order to set a custom template.

File: metacatalog/util/exceptions.py [hunk 5]
```diff
@@ -8,3 +8,7 @@ class MetadataMissingError(RuntimeError):
 
 class IOOperationNotFoundError(RuntimeError, ValueError):
     pass
+
+
+class StandardsExportExtentionNotFoundError(RuntimeError, ValueError):
+    pass
```
Comment (mmaelicke) [inline metacatalog/util/exceptions.py:14]: What do we use this exception for?

Comment (mmaelicke) [inline metacatalog/util/exceptions.py:14]: Also, please add a blank line

File: metacatalog/ext/standard_export/util.py [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+        
+        # autoescape=False: fixes problems with symbol ">" (keyword.path)
+        env = Environment(loader=FileSystemLoader(searchpath=full_path), autoescape=False)
+        
+        # prevent whitespaces / newlines from jinja blocks in template
+        env.trim_blocks = True
+        env.lstrip_blocks = True
+
+        # get template
+        template = env.get_template("iso19115-2.j2")
```
Comment (mmaelicke) [inline metacatalog/ext/standard_export/util.py:36]: The user should have the possibility to choose another template here. 
Also, the path where the jinia2 environment is initialized, should not be hardcoded

File: metacatalog/ext/standard_export/util.py [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+    """
+    # if a base group exists, use the uuid of the base group
+    if rs.group:
+        uuid = rs.group.uuid
+
+    # if no group exists, there is only one member in the ImmutableResultSet -> use its uuid
+    else:
+        uuid = rs.get('uuid')
+
+    return uuid
```
Comment (mmaelicke) [inline metacatalog/ext/standard_export/util.py:61]: Is this still necessary?
The `ImmutableResultSet.uuid` property should already return this, right?

File: metacatalog/ext/standard_export/util.py [hunk 8]
```diff
@@ -0,0 +1,627 @@
...
+                else:
+                    _details[detail_dict['key']] = detail_dict
+                    # remove unwanted key-value pairs
+                    _details[detail_dict['key']] = {key: val for key, val in _details[detail_dict['key']].items() if key in ['value', 'key', 'entry_uuid', 'description']}
+
+            # turn into a transposed dataframe
+            df = pd.DataFrame(_details).T
+
+            # append markdown table to details
+            details.append(df.to_markdown())
```
Comment (mmaelicke) [inline metacatalog/ext/standard_export/util.py:320]: This is not ideal behavior. 
The details should end up as extra metadata info in the iso. It has an entity for that. Just putting tables into the abstract is not the final behavior.

Maybe we keep the function for now, but implement the proper behavior as soon as possible.
Thus, this should open an urgent issue.

File: metacatalog/ext/standard_export/util.py [hunk 9]
```diff
@@ -0,0 +1,627 @@
...
+def _validate_xml(xml: str) -> bool:
+    """
+    Checks whether the input XML is well-formed (correct syntax).
+    Currently, it is not checked whether the input XML is also valid (We could e.g.
+    check with schematron for the 20 mandatory fields).
+    """
+    try: 
+        etree.fromstring(xml)
+    except etree.XMLSyntaxError as e:
+        raise(e)
```
Comment (mmaelicke) [inline metacatalog/ext/standard_export/util.py:627]: please add an empty line at the end of the file.

Also, this function is not returning anything...

**Review by mmaelicke — CHANGES_REQUESTED** [review]
Do we have any plans for unittests?

Should be pretty straightforward to create the etree and check for different values be present? So just a high level unittest that runs the whole extension.

**Comment by mmaelicke** [comment]
@AlexDo1  There is a merge conflict about version numbers. 
There seems to be a version pushed to the iso branch, but I don't get how this can be **higher** than the one in main. iso can't be pulled from a branch that is ahead of main in the first place, so where is version 0.7.7 coming from?
Can't find it on PyPI either. I guess this is a mistake?

**Comment by AlexDo1** [comment]
> @AlexDo1 There is a merge conflict about version numbers. There seems to be a version pushed to the iso branch, but I don't get how this can be **higher** than the one in main. iso can't be pulled from a branch that is ahead of main in the first place, so where is version 0.7.7 coming from? Can't find it on PyPI either. I guess this is a mistake?

Yes, that's a mistake, I changed the version number to `0.7.7` for `versionadded` in the docstrings. I will change the version back to what it was before.

File: metacatalog/ext/standard_export/__init__.py (hunk 1 above)
Comment (AlexDo1) [reply metacatalog/ext/standard_export/__init__.py:1]: But as this is a Class, we should use camel case and no separation between words, right?

**Comment by mmaelicke** [comment]
> Yes, that's a mistake, I changed the version number to `0.7.7` for `versionadded` in the docstrings. I will change the version back to what it was before.

Ok. But only the version. Not the docstrings, as the next version will be 0.7.7

File: metacatalog/util/exceptions.py (hunk 5 above)
Comment (AlexDo1) [reply metacatalog/util/exceptions.py:14]: That's a good question, I used this Exception before, but not anymore.  
I will remove it.

File: metacatalog/ext/standard_export/util.py (hunk 7 above)
Comment (AlexDo1) [reply metacatalog/ext/standard_export/util.py:61]: That's right!

File: metacatalog/ext/standard_export/__init__.py (hunk 1 above)
Comment (mmaelicke) [reply metacatalog/ext/standard_export/__init__.py:1]: Right, but the class has already the right name. Just the File doesn't have the right name.

File: metacatalog/ext/standards_export/test_export.ipynb [hunk 10]
```diff
@@ -0,0 +1,572 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from metacatalog import api, ext\n",
```
Comment (mmaelicke) [inline metacatalog/ext/standards_export/test_export.ipynb:9]: I would like to remove this notebook from the repo and add it to the gitignore. Or is there any specific reason, whiy the user would need it?

**Review by mmaelicke — CHANGES_REQUESTED** [review]
We should at least add a unittest, that will only check if the export to ISO is producing an XML output at all.

File: metacatalog/ext/standards_export/test_export.ipynb (hunk 10 above)
Comment (AlexDo1) [reply metacatalog/ext/standards_export/test_export.ipynb:9]: I pushed that by accident, thanks for pointing this out!

**Comment by AlexDo1** [comment]
Is there a possibility to add to requirements.txt inside an extension, as `jinja2` and `lxml` are only required if the `standards_export` extension is activated.  
And do we want to activate the extension by default?

**Comment by mmaelicke** [comment]
I would not activate it by default.

We  need a import flag that indicates if all requirements were met. The extension activation function should then check if everything is there.
The user has to take care of that.

**Comment by AlexDo1** [comment]
> I would not activate it by default.
> 
> We need a import flag that indicates if all requirements were met. The extension activation function should then check if everything is there. The user has to take care of that.

Yes, but we need to install all required packages to run the unittests for extensions.

**Review by mmaelicke — APPROVED** [review]

**Comment by mmaelicke** [comment]
Nice! This earns the next minor version...

**Pull request merged**